src/Server.zig

```diff
@@ -469,6 +470,101 @@ fn getAstCheckDiagnostics(
     }
 }
 
+/// TODO: Memory
+pub fn getFullCompilationDiagnostics(
+    server: *Server,
+    handle: DocumentStore.Handle,
+    diagnostics: *std.ArrayListUnmanaged(types.Diagnostic),
+) !void {
+    var allocator = server.arena.allocator();
+    const config = server.config;
+
+    const build_file_uri = handle.associated_build_file orelse return;
+    const build_file = server.document_store.build_files.get(build_file_uri).?;
+
+    const tracy_zone = tracy.trace(@src());
+    defer tracy_zone.end();
+
+    var arena = std.heap.ArenaAllocator.init(allocator);
+    defer arena.deinit();
+    const arena_allocator = arena.allocator();
+
+    const build_file_path = try uri_utils.parse(arena_allocator, build_file.uri);
+    const directory_path = try std.fs.path.resolve(arena_allocator, &.{ build_file_path, "../" });
+
+    // TODO extract this option from `BuildAssociatedConfig.BuildOption`
+    const zig_cache_root: []const u8 = try std.fs.path.join(arena_allocator, &.{ directory_path, "zig-cache" });
+    // Since we don't compile anything and no packages should put their
+    // files there this path can be ignored
+    const zig_global_cache_root: []const u8 = "ZLS_DONT_CARE";
+
+    const standard_args = [_][]const u8{
+        config.zig_exe_path.?,
+        "run",
+        config.diagnostics_build_runner_path.?,
+        "--cache-dir",
+        config.global_cache_path.?,
+        "--pkg-begin",
+        "@build@",
+        build_file_path,
+        "--pkg-end",
+        "--",
+        config.zig_exe_path.?,
+        directory_path,
+        zig_cache_root,
+        zig_global_cache_root,
+    };
+
+    const arg_length = standard_args.len + if (build_file.build_associated_config) |cfg| if (cfg.build_options) |options| options.len else 0 else 0;
+    var args = try std.ArrayListUnmanaged([]const u8).initCapacity(arena_allocator, arg_length);
+    args.appendSliceAssumeCapacity(standard_args[0..]);
+    if (build_file.build_associated_config) |cfg| {
+        if (cfg.build_options) |options| {
+            for (options) |opt| {
+                args.appendAssumeCapacity(try opt.formatParam(arena_allocator));
+            }
+        }
+    }
+
+    const zig_run_result = try std.ChildProcess.exec(.{
+        .allocator = arena_allocator,
+        .argv = args.items,
+    });
+
+    var line_iterator = std.mem.split(u8, zig_run_result.stderr, "\n");
```

Comment:
the `\n` character could potentially could be part of a unicode character

Instead, you want to find the first non-ASCII character OR newline. If non-ASCII character, skip ahead by the length of the codepoint. If newline, process accordingly

Here is some example code: https://github.com/oven-sh/bun/blob/b86915c80819e61e6e1932d7802a4c31c73094e9/src/string_immutable.zig#L2919-L2966

Reply:
I think I'll just end up using `std.unicode` like the rest of the codebase, tracked by #872

Reply:
See conversation is Discord for why this mitigation is not necessary